File: src/871. Minimum Number of Refueling Stops.js
/**
 * @param {number} target
 * @param {number} startFuel
 * @param {number[][]} stations
 * @return {number}
 */

export function minRefuelStops(target, startFuel, stations) {
  const n = stations.length,
    dp = Array(n + 1).fill(0)
  dp[0] = startFuel
  for (let i = 0; i < n; i++) {
    for (let t = i; t >= 0; t--) {
      if (dp[t] >= stations[i][0]) {
        dp[t + 1] = Math.max(dp[t + 1], dp[t] + stations[i][1])
      }
    }
  }
  for (let i = 0; i <= n; i++) {
    if (dp[i] >= target) {
      return i
    }
  }
  return -1
}

File: src/871. Minimum Number of Refueling Stops.cpp
// DP
// class Solution {
//     public:
//         int minRefuelStops(int target, int startFuel, vector < vector < int >> & stations) {
//             int n = stations.size();
//             vector < long > dp(n + 1);
//             dp[0] = startFuel;
//             for (int i = 0; i < n; ++i) {
//                 for (int t = i; t >= 0; --t) {
//                     if (dp[t] >= stations[i][0]) {
//                         dp[t + 1] = max(dp[t + 1], dp[t] + stations[i][1]);
//                     }
//                 }
//             }
//             for (int i = 0; i <= n; ++i) {
//                 if (dp[i] >= target) {
//                     return i;
//                 }
//             }
//             return -1;
//         }
// };

// Max Heap
class Solution {
    public:
        int minRefuelStops(int target, int startFuel, vector < vector < int >> & stations) {
            int n = stations.size();
            priority_queue < int > priorityQueue;
            int curr = startFuel;
            int i = 0;
            int result = 0;
            while (curr < target) {
                while (i < n && curr >= stations[i][0]) {
                    priorityQueue.push(stations[i][1]);
                    i++;
                }
                if (priorityQueue.empty()) {
                    return -1;
                }
                int dist = priorityQueue.top();
                priorityQueue.pop();
                curr += dist;
                result++;
            }
            return result;
        }
};